crates/agentgateway/src/mcp/rbac.rs

```diff
@@ -0,0 +1,273 @@
+	Tool(ResourceId),
+	Prompt(ResourceId),
+	Resource(ResourceId),
+}
+
+impl ResourceType {
+	fn entity(&self) -> EntityUid {
+		let (n, i) = match self {
+			ResourceType::Tool(r) => (ENTITY_NAME_TOOL.clone(), EntityId::new(&r.id)),
+			ResourceType::Prompt(r) => (ENTITY_NAME_PROMPT.clone(), EntityId::new(&r.id)),
+			ResourceType::Resource(r) => (ENTITY_NAME_RESOURCE.clone(), EntityId::new(&r.id)),
+		};
+		EntityUid::from_type_name_and_id(n, i)
+	}
+	fn target_entity(&self) -> EntityUid {
+		EntityUid::from_type_name_and_id(ENTITY_NAME_TARGET.clone(), EntityId::new(self.target()))
+	}
+	fn target(&self) -> &str {
+		match self {
+			ResourceType::Tool(r) => &r.target,
+			ResourceType::Prompt(r) => &r.target,
+			ResourceType::Resource(r) => &r.target,
+		}
+	}
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq)]
+#[serde(rename_all = "camelCase")]
+#[cfg_attr(feature = "schema", derive(JsonSchema))]
+pub struct ResourceId {
+	#[serde(default)]
+	target: String,
+	#[serde(default)]
+	id: String,
+}
+
+impl ResourceId {
+	pub fn new(target: String, id: String) -> Self {
+		Self { target, id }
+	}
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct Identity {
+	pub claims: Option<Claims>,
+	pub connection_id: Option<String>,
+}
+
+impl agent_core::trcng::Claim for Identity {
+	fn get_claim(&self, key: &str) -> Option<&str> {
+		self.get_claim(key, ".")
+	}
+}
+
+impl Identity {
+	pub fn empty() -> Self {
+		Self {
+			claims: None,
+			connection_id: None,
+		}
+	}
+
+	pub fn new(claims: Option<Claims>, connection_id: Option<String>) -> Self {
+		Self {
+			claims,
+			connection_id,
+		}
+	}
+	// Attempts to get the claim from the claims map
+	// The key should be split by the key_delimiter and then the map should be searched recursively
+	// If the key is not found, it returns None
+	// If the key is found, it returns the value
+	pub fn get_claim(&self, key: &str, key_delimiter: &str) -> Option<&str> {
+		match &self.claims {
+			Some(claims) => {
+				// Split the key by the delimiter to handle nested lookups
+				let keys = key.split(key_delimiter).collect::<Vec<&str>>();
+
+				// Start with the root claims map
+				let mut current_value = &claims.inner;
+
+				// Navigate through each key level
+				let num_keys = keys.len();
+				for (index, key_part) in keys.into_iter().enumerate() {
+					// Get the value at this level
+					let value = current_value.get(key_part)?;
+
+					// If this is the last key part, return the string value
+					if index == num_keys - 1 {
+						return value.as_str();
+					}
+
+					// Otherwise, try to navigate deeper if it's an object
+					current_value = value.as_object()?;
+				}
+
+				None
+			},
+			None => None,
+		}
+	}
+}
+
+lazy_static! {
+	static ref PRINCIPAL_UID_ANONYMOUS: EntityUid = "AnonymousUser::\"\"".parse().unwrap();
+	static ref PRINCIPAL_NAME_USER: EntityTypeName = "User".parse().unwrap();
+	static ref ACTION_CALL_TOOL: EntityUid = r#"Action::"call_tool""#.parse().unwrap();
```

Comment:
So "Action" call_tool makes sense for specifying policy around tools. But It seems we have other resource targets beyond tools, like Prompt:: and Resource:: ? Should this be "call_tool" or just "call" ?.

Reply:
So we have

get/list prompt
read/list resource
call/list tools

maybe our action types are "read", "call", and "list"?